embedded/sql/stmt.go

```diff
@@ -487,53 +452,43 @@ func (stmt *UpsertIntoStmt) inferParameters(e *Engine, implicitDB *Database, par
 	return nil
 }
 
-func (stmt *UpsertIntoStmt) validate(table *Table) (map[uint64]int, error) {
-	pkIncluded := false
-	selByColID := make(map[uint64]int, len(stmt.cols))
+func (stmt *UpsertIntoStmt) validate(table *Table) (map[uint32]int, error) {
+	selPosByColID := make(map[uint32]int, len(stmt.cols))
 
 	for i, c := range stmt.cols {
 		col, err := table.GetColumnByName(c)
 		if err != nil {
 			return nil, err
 		}
 
-		if table.pk.colName == c {
-			pkIncluded = true
-		}
-
-		_, duplicated := selByColID[col.id]
+		_, duplicated := selPosByColID[col.id]
 		if duplicated {
 			return nil, ErrDuplicatedColumn
 		}
 
-		selByColID[col.id] = i
+		selPosByColID[col.id] = i
 	}
 
-	if !pkIncluded && (!stmt.isInsert || !table.pk.autoIncrement) {
-		return nil, ErrPKCanNotBeNull
-	}
-
-	if pkIncluded && stmt.isInsert && table.pk.autoIncrement {
-		return nil, ErrNoValueForAutoIncrementalColumn
-	}
-
-	return selByColID, nil
+	return selPosByColID, nil
 }
 
 func (stmt *UpsertIntoStmt) compileUsing(e *Engine, implicitDB *Database, params map[string]interface{}) (summary *TxSummary, err error) {
-	summary = newTxSummary()
-	summary.db = implicitDB
-
 	if implicitDB == nil {
 		return nil, ErrNoDatabaseSelected
 	}
 
+	summary = newTxSummary(implicitDB)
+
 	table, err := stmt.tableRef.referencedTable(e, implicitDB)
 	if err != nil {
 		return nil, err
 	}
 
-	cs, err := stmt.validate(table)
+	if !stmt.isInsert && len(table.indexes) > 1 {
+		return nil, ErrLimitedUpsert
```

Comment:
Isn't that only about unique indexes?

Reply:
actually not, the issue was before this PR, because the index entry with the old value will still be present in the index